userspace/libsinsp/event.cpp

```diff
@@ -2084,76 +2088,69 @@ const char* sinsp_evt::get_param_as_str(uint32_t id, OUT const char** resolved_s
 	case PT_ENUMFLAGS16:
 	case PT_ENUMFLAGS32:
 		{
-			uint32_t val = 0;
-			switch(param_info->type)
-			{
-			case PT_FLAGS8:
-			case PT_ENUMFLAGS8:
-				val = *(uint8_t *)payload;
-				break;
-			case PT_FLAGS16:
-			case PT_ENUMFLAGS16:
-				val = *(uint16_t *)payload;
-				break;
-			case PT_FLAGS32:
-			case PT_ENUMFLAGS32:
-				val = *(uint32_t *)payload;
-				break;
-			default:
-				ASSERT(false);
-			}
+			uint32_t val = *(uint32_t *)payload & (((uint64_t)1 << payload_len * 8) - 1);
 			snprintf(&m_paramstr_storage[0],
 				     m_paramstr_storage.size(),
 				     "%" PRIu32, val);
 
-			const bool exact_match = param_info->type == PT_ENUMFLAGS8 || param_info->type == PT_ENUMFLAGS16 || param_info->type == PT_ENUMFLAGS32;
-
 			const struct ppm_name_value *flags = (const struct ppm_name_value *)m_info->params[id].info;
+			const bool exact_match = param_info->type == PT_ENUMFLAGS8 || param_info->type == PT_ENUMFLAGS16 || param_info->type == PT_ENUMFLAGS32;
 			const char *separator = "";
 			uint32_t initial_val = val;
 			uint32_t j = 0;
 
-			while(flags != NULL && flags->name != NULL && flags->value != initial_val)
+			while(flags != NULL && flags->name != NULL)
 			{
-				// If flag is 0, then initial_val needs to be 0 for the flag to be resolved
-				if ((exact_match && flags->value == initial_val) ||
-				    ((flags->value == 0 && initial_val == 0) ||
-				    (flags->value != 0 && (val & flags->value) == flags->value && val != 0)))
+				bool match = false;
+				if (exact_match)
+				{
+					match = flags->value == initial_val;
+				}
+				else
+				{
+					// If flag is 0, then initial_val needs to be 0 for the flag to be resolved
+					if ((flags->value == 0 && initial_val == 0) ||
+					   (flags->value != 0 && (val & flags->value) == flags->value && val != 0))
+					{
+						match = true;
+						// We remove current flags value to avoid duplicate flags e.g. PPM_O_RDWR, PPM_O_RDONLY, PPM_O_WRONLY
+						val &= ~flags->value;
+					}
+				}
+				if (match)
 				{
 					if(m_resolved_paramstr_storage.size() < j + strlen(separator) + strlen(flags->name))
 					{
 						m_resolved_paramstr_storage.resize(m_resolved_paramstr_storage.size() * 2);
 					}
 
 					j += snprintf(&m_resolved_paramstr_storage[j],
-								  m_resolved_paramstr_storage.size(),
-							 	  "%s%s",
-							 	  separator,
-							 	  flags->name);
-
+						      m_resolved_paramstr_storage.size(),
+						      "%s%s",
+						      separator,
+						      flags->name);
 					if (exact_match)
 					{
-						// We found the match!
-						break;
+						// Multiple flags name match same enum value
+						separator = " ";
```

Comment:
We use just a space as a separator for multiple flags name matching same enum value.

Reply:
butwhy.gif

Reply:
Because it makes no sense for enum flags to use `|`. We could use a comma perhaps.

Reply:
WDYT?

Reply:
@gnosek :)

Reply:
Yep i tested, `|` is best. Fixed up, thank you!